server/routerlicious/packages/services-shared/src/webServer.ts

```diff
@@ -106,3 +111,236 @@ export class BasicWebServerFactory implements core.IWebServerFactory {
+	): http.Server {
+		// Init Primary cluster thread
+		Lumberjack.info(`Primary cluster process is running.`, { pid: process.pid });
+
+		// Create a blank HTTP Server that will distribute incoming requests to worker nodes.
+		const server = createAndConfigureHttpServer(undefined, undefined);
+		additionalServerSetup?.(server);
+
+		// Regularly kill stuck workers
+		const heartbeatTimeoutMs =
+			(this.clusterConfig?.workerTimeoutNumMissedHeartbeats ?? 3) * this.heartbeatIntervalMs;
+		setInterval(() => {
+			for (const [workerId, lastHeartbeat] of this.lastHeartbeatMap.entries()) {
+				const msSinceLastHeartbeat = Date.now() - lastHeartbeat;
+				if (msSinceLastHeartbeat >= heartbeatTimeoutMs) {
+					const worker = cluster.workers?.[workerId];
+					if (!worker) {
+						Lumberjack.error("Could not find worker after heartbeat timeout.", {
+							id: workerId,
+						});
+						return;
+					}
+					Lumberjack.error(`Worker heartbeat timeout.`, {
+						pid: worker?.process.pid,
+						msSinceLastHeartbeat,
+					});
+					this.killWorker(worker);
+				}
+			}
+		}, heartbeatTimeoutMs);
+
+		// Spawn initial number of workers according to configs or available CPUs
+		const numWorkers = this.clusterConfig?.numMaxWorkers ?? availableParallelism();
+		Lumberjack.info(`Spawning ${numWorkers} cluster workers.`);
+		for (let i = 0; i < numWorkers; i++) {
+			this.spawnWorker();
+		}
+
+		// Kill workers when they take too long to spawn.
+		cluster.on("fork", (worker) => {
+			const timeout = setTimeout(
+				() => {
+					Lumberjack.error("Timed out waiting for worker to spawn.", {
+						id: worker.id,
+						pid: worker.process.pid,
+					});
+
+					// Make sure worker dies.
+					this.killWorker(worker);
+
+					// Remove timeout from map.
+					this.newForkTimeouts.delete(worker.id);
+				},
+				this.clusterConfig?.workerForkTimeoutMs,
+			);
+			this.newForkTimeouts.set(worker.id, timeout);
+		});
+		cluster.on("listening", (worker, address) => {
+			clearTimeout(this.newForkTimeouts.get(worker.id));
+			this.newForkTimeouts.delete(worker.id);
+		});
+
+		// Listen for exiting workers.
+		cluster.on("exit", (worker, code, signal) => {
+			const logLevel = signal || code !== 0 ? "error" : "info";
+			Lumberjack[logLevel](`Worker died.`, {
+				code,
+				signal,
+				pid: worker.process.pid,
+				id: worker.id,
+			});
+
+			// Clear timeout trackers
+			this.lastHeartbeatMap.delete(worker.id);
+			clearTimeout(this.newForkTimeouts.get(worker.id));
+			this.newForkTimeouts.delete(worker.id);
+			clearTimeout(this.disconnectTimeouts.get(worker.id));
+			this.disconnectTimeouts.delete(worker.id);
+
+			// Spawn a new worker to replace dead worker if died due to error.
+			if (logLevel === "error") {
+				Lumberjack.info(`Spawning new worker to replace dead worker.`);
+				this.spawnWorker();
+			}
+		});
+
+		return server;
+	}
+
+	protected initializeWorkerThread(requestListener: RequestListener): http.Server {
```

Comment:
Now that we would be running some sort of multithreaded web server, is there any potential race condition that could arise? For example, are there shared resources used by different threads?

Reply:
Not that I can think of. You can kinda think of each of these threads as a separate service pod.